text/0000-edition-2021.md

```diff
@@ -0,0 +1,238 @@
+
+## Editions do not split the ecosystem
+
+The most important rule for editions is that crates in one edition can interoperate seamlessly with crates compiled in other editions. This ensures that the decision to migrate to a newer edition is a "private one" that the crate can make without affecting others, apart from the fact that it affects the version of rustc that is required, akin to making use of any new feature.
+
+The requirement for crate interoperability implies some limits on the kinds of changes that we can make in an edition. In general, changes that occur in an edition tend to be "skin deep". All Rust code, regardless of edition, is ultimately compiled to the same internal representation within the compiler.
+
+## Edition migration is easy and largely automated
+
+Our goal is to make it easy for crates to upgrade to newer editions. Whenever we release a new edition, we also release tooling to automate the migration. The tooling is not necessarily perfect: it may not cover all corner cases, and manual changes may still be required. The tooling tries hard to avoid changes to semantics that could affect the correctness or performance of the code.
+
+In addition to tooling, we also maintain an Edition Migration Guide that covers the changes that are part of an edition. This guide will describe the change and give pointers to where people can learn more about it. It will also cover any corner cases or details that people should be aware of. The guide serves both as an overview of the edition, but also as a quick troubleshooting reference if people encounter problems with the automated tooling.
+
+## Users control when they adopt the new edition
+
+Part of making edition migration easy is ensuring that users can choose when they wish to upgrade. We recognize that many users, particularly production users, will need to schedule time to manage an Edition upgrade as part of their overall development cycle. 
+
+We also want to enable upgrading to be done in a gradual fashion, meaning that it is possible to migrate a crate module-by-module in separate steps, with the final move to the new edition happening only when all modules are migrated. This is done by ensuring that there is always an "intersection" that is compatible with both the edition N and its successor N+1. 
+
+## Editions are meant to be adopted
+
+Our goal is to see all Rust users adopt the new edition, just as we would generally prefer for people to move to the "latest and greatest" ways of using Rust once they are available. We try primarily to achieve this by "soft means", such as changing the default of cargo to the latest edition, or making new features that entice people to move. However, we may sometimes opt to issue warnings or deprecations where the best fix involves moving to a new edition.
```

Comment:
Hmm. I think the language of "entice people to move" was confusing and not really needed. I didn't really mean that we *actively try to entice* people, just that as we design new features, people will hopefully adopt the new edition because they want them. Maybe I can just remove that phrase.

Reply:
Yeah, I also find it odd and I would go as far as recommending widely-used crates to withhold migrating for quite some time. (See my concerns in other comment.)

Maybe phrase it as: "People who can afford to use the compiler supporting the new edition are encouraged to migrate to the new edition."

Reply:
I tweaked the language and confirmed with @joshtriplett that he was happy with it now.